src/main/java/uk/gov/justice/digital/delius/service/ReferralService.java

```diff
@@ -1,46 +1,85 @@
-            .header(HttpHeaders.ACCEPT, MediaType.APPLICATION_JSON_VALUE)
-            .bodyValue(contact)
-            .retrieve()
-            .toEntity(String.class)
-            .block();
+            .eventId(referralSent.getConvictionId())
+            .requirementId(referralSent.getRequirementId())
+            .referralDate(referralSent.getDate())
+            .status(referralSent.getNsiStatus())
+            .statusDate(referralSent.getDate().atStartOfDay())
+            .notes(referralSent.getNotes())
+            .intendedProvider(referralSent.getProviderCode())
+            .manager(NewNsiManager.builder()
+                .staff(referralSent.getStaffCode())
+                .team(referralSent.getTeamCode())
+                .provider(referralSent.getProviderCode())
+                .build()
+            ).build();
+
+        var newNsiReponse = deliusApiClient.createNewNsi(newNsiRequest);
+
+        return ReferralSentResponse.builder().nsiId(newNsiReponse.getId()).build();
+    }
+
+    public Optional<Nsi> getExistingMatchingNsi(String crn, ReferralSentRequest referralSent) {
+        // determine if there is an existing suitable NSI
+        var offenderId = offenderService.offenderIdOfCrn(crn).orElseThrow();
+
+        var existingNsis = nsiService.getNsiByCodes(offenderId, referralSent.getConvictionId(), Collections.singletonList(referralSent.getNsiType()))
+            .map(wrapper -> wrapper.getNsis().stream()
+                // eventID, offenderID, nsiID, and callerID are handled in the NSI service
+                .filter(nsi -> Optional.ofNullable(nsi.getNsiSubType()).map(n -> n.getCode().equals(referralSent.getNsiSubType())).orElse(false))
+                .filter(nsi -> Optional.ofNullable(nsi.getReferralDate()).map(n -> n.equals(referralSent.getDate())).orElse(false))
+                .filter(nsi -> Optional.ofNullable(nsi.getNsiStatus()).map(n -> n.getCode().equals(referralSent.getNsiStatus())).orElse(false))
+                .filter(nsi -> Optional.ofNullable(nsi.getRequirement()).map(n -> nsi.getRequirement().getRequirementId().equals(referralSent.getRequirementId())).orElse(false))
+                .filter(nsi -> Optional.ofNullable(nsi.getIntendedProvider()).map(n -> n.getCode().equals(referralSent.getProviderCode())).orElse(false))
+                .filter(nsi -> Optional.ofNullable(nsi.getNsiManagers()).map(n -> n.stream().anyMatch(nsiManager -> nsiManager.getStaff().getStaffCode().equals(referralSent.getStaffCode())
+                    && nsiManager.getProbationArea().getCode().equals(referralSent.getProviderCode())
+                    && nsiManager.getTeam().getCode().equals(referralSent.getTeamCode()))).orElse(false))
+                .collect(toList())
+            ).orElseThrow(() -> new RuntimeException("Offender ID not found"));
+
+        if(existingNsis.size() > 1) {
+            throw new RuntimeException("Multiple existing matching NSIs found");
```

Comment:
maybe some other exception that would return a 409 (conflict) else this will be a 500 where as 409 indicates we have conflicting data that can not be resolved